File: java/src/main/java/io/github/nickliapin/tdc/packs/Json.java
package io.github.nickliapin.tdc.packs;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * A small, strict JSON reader — enough to read a config file, and no more.
 *
 * <p>Written rather than depended on for the same reason the Parquet writer is: this library
 * ships with no runtime dependencies, and one added to read a twenty-line settings file would be
 * a poor trade. Strict on purpose — a file the reference rejects has to be rejected here too, or
 * the two disagree about which projects are configured correctly.
 */
public final class Json {

  /** Anything malformed, reported with where it went wrong. */
  public static final class SyntaxException extends RuntimeException {

    private static final long serialVersionUID = 1L;

    SyntaxException(String message) {
      super(message);
    }
  }

  private final String text;
  private int at;

  private Json(String text) {
    this.text = text;
  }

  /**
   * Parse a document.
   *
   * @return a {@code Map}, {@code List}, {@code String}, {@code Double}, {@code Boolean}, or
   *     {@code null}
   */
  public static Object parse(String text) {
    Json reader = new Json(text);
    reader.skipSpace();
    Object value = reader.value();
    reader.skipSpace();
    if (reader.at < reader.text.length()) {
      throw new SyntaxException("unexpected trailing text at position " + reader.at);
    }
    return value;
  }

  private Object value() {
    if (at >= text.length()) {
      throw new SyntaxException("unexpected end of input");
    }
    char c = text.charAt(at);
    switch (c) {
      case '{':
        return object();
      case '[':
        return array();
      case '"':
        return string();
      case 't':
        return literal("true", Boolean.TRUE);
      case 'f':
        return literal("false", Boolean.FALSE);
      case 'n':
        return literal("null", null);
      default:
        return number();
    }
  }

  private Map<String, Object> object() {
    Map<String, Object> out = new LinkedHashMap<>();
    at++; // '{'
    skipSpace();
    if (peek() == '}') {
      at++;
      return out;
    }
    while (true) {
      skipSpace();
      if (peek() != '"') {
        throw new SyntaxException("expected a key at position " + at);
      }
      String key = string();
      skipSpace();
      expect(':');
      skipSpace();
      out.put(key, value());
      skipSpace();
      char c = next();
      if (c == '}') {
        return out;
      }
      if (c != ',') {
        throw new SyntaxException("expected ',' or '}' at position " + (at - 1));
      }
    }
  }

  private List<Object> array() {
    List<Object> out = new ArrayList<>();
    at++; // '['
    skipSpace();
    if (peek() == ']') {
      at++;
      return out;
    }
    while (true) {
      skipSpace();
      out.add(value());
      skipSpace();
      char c = next();
      if (c == ']') {
        return out;
      }
      if (c != ',') {
        throw new SyntaxException("expected ',' or ']' at position " + (at - 1));
      }
    }
  }

  private String string() {
    at++; // opening quote
    StringBuilder out = new StringBuilder();
    while (true) {
      if (at >= text.length()) {
        throw new SyntaxException("unterminated string");
      }
      char c = text.charAt(at++);
      if (c == '"') {
        return out.toString();
      }
      if (c != '\\') {
        out.append(c);
        continue;
      }
      char escape = next();
      switch (escape) {
        case '"' -> out.append('"');
        case '\\' -> out.append('\\');
        case '/' -> out.append('/');
        case 'b' -> out.append('\b');
        case 'f' -> out.append('\f');
        case 'n' -> out.append('\n');
        case 'r' -> out.append('\r');
        case 't' -> out.append('\t');
        case 'u' -> {
          if (at + 4 > text.length()) {
            throw new SyntaxException("truncated \\u escape");
          }
          out.append((char) Integer.parseInt(text.substring(at, at + 4), 16));
          at += 4;
        }
        default -> throw new SyntaxException("unknown escape \\" + escape);
      }
    }
  }

  private Double number() {
    int start = at;
    if (peek() == '-') {
      at++;
    }
    while (at < text.length() && "0123456789.eE+-".indexOf(text.charAt(at)) >= 0) {
      at++;
    }
    if (start == at) {
      throw new SyntaxException("unexpected character '" + text.charAt(at) + "' at position " + at);
    }
    try {
      return Double.valueOf(text.substring(start, at));
    } catch (NumberFormatException e) {
      throw new SyntaxException("not a number: \"" + text.substring(start, at) + "\"");
    }
  }

  private Object literal(String word, Object value) {
    if (!text.startsWith(word, at)) {
      throw new SyntaxException("unexpected value at position " + at);
    }
    at += word.length();
    return value;
  }

  private void skipSpace() {
    while (at < text.length() && Character.isWhitespace(text.charAt(at))) {
      at++;
    }
  }

  private char peek() {
    if (at >= text.length()) {
      throw new SyntaxException("unexpected end of input");
    }
    return text.charAt(at);
  }

  private char next() {
    char c = peek();
    at++;
    return c;
  }

  private void expect(char c) {
    if (next() != c) {
      throw new SyntaxException("expected '" + c + "' at position " + (at - 1));
    }
  }
}

File: rust/src/json.rs
//! Just enough JSON, written here rather than depended on.
//!
//! Rust has no JSON in its standard library, and the other four implementations
//! all had one to hand. Three reasons this is a module and not a dependency:
//! the crate stays dependency-free, so `cargo add tdcv2` pulls nothing and
//! builds offline; the shared fixtures are read by the tests through this same
//! module, so every test exercises it; and the shapes involved are a config file
//! and a registry index, not arbitrary documents.
//!
//! An object keeps its keys **in order**, in a `Vec` rather than a map. That is
//! not an oversight: `tdcv2 pack add` rewrites the project's config file, and it
//! must give back every key it found — including ones this version has never
//! heard of — in the order they were written. A hash map would silently reorder
//! a user's file on every install, and one shared fixture case exists precisely
//! to catch that.

use std::collections::BTreeMap;
use std::fmt::Write as _;

#[derive(Clone, Debug, PartialEq)]
pub enum Value {
    Null,
    Bool(bool),
    Number(f64),
    String(String),
    Array(Vec<Value>),
    /// Key/value pairs in source order. Duplicate keys keep the last value, as
    /// every other implementation's parser does.
    Object(Vec<(String, Value)>),
}

#[derive(Debug)]
pub struct JsonError {
    pub message: String,
    /// 0-based byte offset where parsing stopped — enough to say "near here".
    pub offset: usize,
}

impl std::fmt::Display for JsonError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} (at byte {})", self.message, self.offset)
    }
}

impl std::error::Error for JsonError {}

impl Value {
    pub fn as_str(&self) -> Option<&str> {
        match self {
            Value::String(s) => Some(s),
            _ => None,
        }
    }

    pub fn as_f64(&self) -> Option<f64> {
        match self {
            Value::Number(n) => Some(*n),
            _ => None,
        }
    }

    pub fn as_i64(&self) -> Option<i64> {
        self.as_f64().map(|n| n as i64)
    }

    pub fn as_bool(&self) -> Option<bool> {
        match self {
            Value::Bool(b) => Some(*b),
            _ => None,
        }
    }

    pub fn as_array(&self) -> Option<&[Value]> {
        match self {
            Value::Array(items) => Some(items),
            _ => None,
        }
    }

    pub fn as_object(&self) -> Option<&[(String, Value)]> {
        match self {
            Value::Object(entries) => Some(entries),
            _ => None,
        }
    }

    /// One member of an object, or `None` for anything else.
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.as_object()?
            .iter()
            .rev()
            .find(|(k, _)| k == key)
            .map(|(_, v)| v)
    }

    /// An object read as a string map — for fixtures whose keys are data.
    pub fn as_string_map(&self) -> BTreeMap<String, String> {
        self.as_object()
            .unwrap_or(&[])
            .iter()
            .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
            .collect()
    }
}

pub fn parse(text: &str) -> Result<Value, JsonError> {
    let bytes = text.as_bytes();
    let mut p = Parser { bytes, pos: 0 };
    p.skip_ws();
    let value = p.value()?;
    p.skip_ws();
    if p.pos != bytes.len() {
        return Err(p.err("trailing characters after the top-level value"));
    }
    Ok(value)
}

struct Parser<'a> {
    bytes: &'a [u8],
    pos: usize,
}

impl<'a> Parser<'a> {
    fn err(&self, message: &str) -> JsonError {
        JsonError {
            message: message.to_string(),
            offset: self.pos,
        }
    }

    fn peek(&self) -> Option<u8> {
        self.bytes.get(self.pos).copied()
    }

    fn skip_ws(&mut self) {
        while matches!(self.peek(), Some(b' ' | b'\t' | b'\r' | b'\n')) {
            self.pos += 1;
        }
    }

    fn expect(&mut self, b: u8) -> Result<(), JsonError> {
        if self.peek() == Some(b) {
            self.pos += 1;
            Ok(())
        } else {
            Err(self.err(&format!("expected '{}'", b as char)))
        }
    }

    fn literal(&mut self, word: &str, value: Value) -> Result<Value, JsonError> {
        if self.bytes[self.pos..].starts_with(word.as_bytes()) {
            self.pos += word.len();
            Ok(value)
        } else {
            Err(self.err("not a valid JSON value"))
        }
    }

    fn value(&mut self) -> Result<Value, JsonError> {
        match self.peek() {
            Some(b'{') => self.object(),
            Some(b'[') => self.array(),
            Some(b'"') => Ok(Value::String(self.string()?)),
            Some(b't') => self.literal("true", Value::Bool(true)),
            Some(b'f') => self.literal("false", Value::Bool(false)),
            Some(b'n') => self.literal("null", Value::Null),
            Some(_) => self.number(),
            None => Err(self.err("unexpected end of input")),
        }
    }

    fn object(&mut self) -> Result<Value, JsonError> {
        self.expect(b'{')?;
        let mut entries: Vec<(String, Value)> = Vec::new();
        self.skip_ws();
        if self.peek() == Some(b'}') {
            self.pos += 1;
            return Ok(Value::Object(entries));
        }
        loop {
            self.skip_ws();
            let key = self.string()?;
            self.skip_ws();
            self.expect(b':')?;
            self.skip_ws();
            let value = self.value()?;
            entries.push((key, value));
            self.skip_ws();
            match self.peek() {
                Some(b',') => self.pos += 1,
                Some(b'}') => {
                    self.pos += 1;
                    return Ok(Value::Object(entries));
                }
                _ => return Err(self.err("expected ',' or '}' in an object")),
            }
        }
    }

    fn array(&mut self) -> Result<Value, JsonError> {
        self.expect(b'[')?;
        let mut items = Vec::new();
        self.skip_ws();
        if self.peek() == Some(b']') {
            self.pos += 1;
            return Ok(Value::Array(items));
        }
        loop {
            self.skip_ws();
            items.push(self.value()?);
            self.skip_ws();
            match self.peek() {
                Some(b',') => self.pos += 1,
                Some(b']') => {
                    self.pos += 1;
                    return Ok(Value::Array(items));
                }
                _ => return Err(self.err("expected ',' or ']' in an array")),
            }
        }
    }

    fn string(&mut self) -> Result<String, JsonError> {
        self.expect(b'"')?;
        let mut out = String::new();
        loop {
            let b = self.peek().ok_or_else(|| self.err("unterminated string"))?;
            self.pos += 1;
            match b {
                b'"' => return Ok(out),
                b'\\' => {
                    let esc = self.peek().ok_or_else(|| self.err("unterminated escape"))?;
                    self.pos += 1;
                    match esc {
                        b'"' => out.push('"'),
                        b'\\' => out.push('\\'),
                        b'/' => out.push('/'),
                        b'b' => out.push('\u{8}'),
                        b'f' => out.push('\u{c}'),
                        b'n' => out.push('\n'),
                        b'r' => out.push('\r'),
                        b't' => out.push('\t'),
                        b'u' => out.push(self.unicode_escape()?),
                        _ => return Err(self.err("unknown escape sequence")),
                    }
                }
                _ => {
                    // Anything else is copied through as-is. The input is a
                    // `&str`, so a multi-byte character arrives as its UTF-8
                    // bytes and pushing them one at a time would not be valid
                    // UTF-8 — so the whole character is taken at once.
                    let start = self.pos - 1;
                    let mut end = self.pos;
                    while end < self.bytes.len() && (self.bytes[end] & 0xC0) == 0x80 {
                        end += 1;
                    }
                    self.pos = end;
                    match std::str::from_utf8(&self.bytes[start..end]) {
                        Ok(s) => out.push_str(s),
                        Err(_) => return Err(self.err("invalid UTF-8 in a string")),
                    }
                }
            }
        }
    }

    /// `\uXXXX`, including the surrogate pair a character above U+FFFF needs.
    fn unicode_escape(&mut self) -> Result<char, JsonError> {
        let first = self.hex4()?;
        if !(0xD800..0xDC00).contains(&first) {
            return char::from_u32(u32::from(first)).ok_or_else(|| self.err("invalid code point"));
        }
        // A high surrogate is only half a character; the low half must follow.
        if !self.bytes[self.pos..].starts_with(b"\\u") {
            return Err(self.err("lone high surrogate in a string"));
        }
        self.pos += 2;
        let second = self.hex4()?;
        if !(0xDC00..0xE000).contains(&second) {
            return Err(self.err("high surrogate not followed by a low one"));
        }
        let combined = 0x10000 + ((u32::from(first) - 0xD800) << 10) + (u32::from(second) - 0xDC00);
        char::from_u32(combined).ok_or_else(|| self.err("invalid code point"))
    }

    fn hex4(&mut self) -> Result<u16, JsonError> {
        if self.pos + 4 > self.bytes.len() {
            return Err(self.err("truncated \\u escape"));
        }
        let text = std::str::from_utf8(&self.bytes[self.pos..self.pos + 4])
            .map_err(|_| self.err("invalid \\u escape"))?;
        let value = u16::from_str_radix(text, 16).map_err(|_| self.err("invalid \\u escape"))?;
        self.pos += 4;
        Ok(value)
    }

    fn number(&mut self) -> Result<Value, JsonError> {
        let start = self.pos;
        if self.peek() == Some(b'-') {
            self.pos += 1;
        }
        while matches!(
            self.peek(),
            Some(b'0'..=b'9' | b'.' | b'e' | b'E' | b'+' | b'-')
        ) {
            self.pos += 1;
        }
        let text = std::str::from_utf8(&self.bytes[start..self.pos])
            .map_err(|_| self.err("invalid number"))?;
        text.parse::<f64>()
            .map(Value::Number)
            .map_err(|_| JsonError {
                message: format!("invalid number \"{text}\""),
                offset: start,
            })
    }
}

/// Write a value back out, two-space indented and newline-terminated.
///
/// The shape every implementation writes for `tdcv2.config.json`, because one
/// shared fixture case compares the file byte for byte.
pub fn to_pretty(value: &Value) -> String {
    let mut out = String::new();
    write_value(&mut out, value, 0);
    out.push('\n');
    out
}

fn write_value(out: &mut String, value: &Value, indent: usize) {
    match value {
        Value::Null => out.push_str("null"),
        Value::Bool(true) => out.push_str("true"),
        Value::Bool(false) => out.push_str("false"),
        Value::Number(n) => out.push_str(&format_number(*n)),
        Value::String(s) => write_string(out, s),
        Value::Array(items) if items.is_empty() => out.push_str("[]"),
        Value::Array(items) => {
            out.push_str("[\n");
            for (i, item) in items.iter().enumerate() {
                pad(out, indent + 1);
                write_value(out, item, indent + 1);
                if i + 1 < items.len() {
                    out.push(',');
                }
                out.push('\n');
            }
            pad(out, indent);
            out.push(']');
        }
        Value::Object(entries) if entries.is_empty() => out.push_str("{}"),
        Value::Object(entries) => {
            out.push_str("{\n");
            for (i, (key, item)) in entries.iter().enumerate() {
                pad(out, indent + 1);
                write_string(out, key);
                out.push_str(": ");
                write_value(out, item, indent + 1);
                if i + 1 < entries.len() {
                    out.push(',');
                }
                out.push('\n');
            }
            pad(out, indent);
            out.push('}');
        }
    }
}

fn pad(out: &mut String, indent: usize) {
    for _ in 0..indent {
        out.push_str("  ");
    }
}

/// An integral double prints without a `.0`, the way every other JSON writer does.
fn format_number(n: f64) -> String {
    if n.is_finite() && n.fract() == 0.0 && n.abs() < 1e15 {
        format!("{}", n as i64)
    } else {
        format!("{n}")
    }
}

fn write_string(out: &mut String, s: &str) {
    out.push('"');
    for ch in s.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if (c as u32) < 0x20 => {
                let _ = write!(out, "\\u{:04x}", c as u32);
            }
            c => out.push(c),
        }
    }
    out.push('"');
}
